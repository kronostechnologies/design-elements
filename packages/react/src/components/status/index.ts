export { Status } from './status';
export type { StatusProps, StatusType } from './status';
