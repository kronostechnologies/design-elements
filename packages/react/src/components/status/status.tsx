import React, { VoidFunctionComponent } from 'react';
import styled from 'styled-components';
import { Theme } from '../../themes';

export type StatusType = 'enabled' | 'disabled' | 'blocked';

function getBackgroundColor(type: StatusType, theme: Theme): string {
    switch (type) {
        case 'enabled':
            return theme.notifications['success-1.1'];
        case 'disabled':
            return theme.greys.white;
        case 'blocked':
            return theme.notifications['error-2.1'];
    }
}

const Wrapper = styled.div<{ type: StatusType }>`
    align-items: center;
    display: flex;

    ${({ type, theme }) => type === 'disabled' && `color: ${theme.greys['dark-grey']}`}
`;

const Circle = styled.div<{ type: StatusType }>`
    background-color: ${({ type, theme }) => getBackgroundColor(type, theme)};
    border: ${({ type, theme }) => (type === 'disabled' ? `1px solid ${theme.greys['dark-grey']}` : 'none')};
    border-radius: 50%;
    box-sizing: border-box;
    height: 10px;
    margin-right: calc(var(--spacing-base) * 1.5);
    width: 10px;
`;

interface Props {
    className?: string;
    label: string;
    type: StatusType;
}

export const Status: VoidFunctionComponent<Props> = ({ className, label, type }) => (
    <Wrapper className={className} type={type}>
        <Circle type={type} />
        <span>{label}</span>
    </Wrapper>
);
