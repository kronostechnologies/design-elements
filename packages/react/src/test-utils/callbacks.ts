export function doNothing(): void {
    // do nothing
}
