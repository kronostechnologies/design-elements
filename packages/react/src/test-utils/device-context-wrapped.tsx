import React, { ReactElement } from 'react';
import { DeviceContextProvider, DeviceType } from '../components/device-context-provider/device-context-provider';

export function DeviceContextWrapped(children: ReactElement, device?: DeviceType): ReactElement {
    return (
        <DeviceContextProvider staticDevice={device}>
            {children}
        </DeviceContextProvider>
    );
}
