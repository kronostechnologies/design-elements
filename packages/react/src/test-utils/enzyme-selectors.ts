import { ReactWrapper, ShallowWrapper } from 'enzyme';

enum ModifierEnum {
    StartsWith = '^',
    EndsWith = '$',
    ContainsOptionalHyphen = '|',
    Contains = '*',
    ContainsWord = '~',
    None = '',
}
type SelectorModifier = `${ModifierEnum}`;

function filterHtmlNodesOnly<W extends ReactWrapper | ShallowWrapper>(
    nodes: W,
): W {
    return nodes
        .hostNodes()
        .filterWhere((x) => x.name() !== null && !x.name().startsWith('styled.')) as W;
}

export function findByTestId<W extends ReactWrapper | ShallowWrapper>(
    wrapper: W,
    testId: string,
    options?: {
        modifier?: SelectorModifier,
        nodeName?: string,
        // Don't use with shallow wrapper as it contains styled node over the html nodes.
        htmlNodesOnly?: boolean,
    },
): W {
    const modifier = options?.modifier ?? '';

    let nodes = wrapper.find(`[data-testid${modifier}="${testId}"]`) as W;

    if (options?.nodeName) {
        nodes = nodes.filterWhere((x) => x.name() === options.nodeName) as W;
    }
    if (options?.htmlNodesOnly) {
        nodes = filterHtmlNodesOnly(nodes);
    }

    return nodes;
}

export function getByTestId<W extends ReactWrapper | ShallowWrapper>(
    wrapper: W,
    testId: string,
    options?: {
        modifier?: SelectorModifier,
        nodeName?: string,
        // Don't use with shallow wrapper as it contains styled node instead of the html nodes.
        htmlNodesOnly?: boolean,
    },
): W {
    return findByTestId(wrapper, testId, options).at(0) as W;
}
