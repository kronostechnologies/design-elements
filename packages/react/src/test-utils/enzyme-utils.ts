import { ReactWrapper } from 'enzyme';
import { act } from 'react-dom/test-utils';

export async function waitForComponentToPaint<P = {}>(
    wrapper: ReactWrapper<P>,
    amount = 0,
): Promise<void> {
    await act(async () => {
        await new Promise((resolve) => setTimeout(resolve, amount));
        wrapper.update();
    });
}

export function expectFocusToBeOn(element: ReactWrapper): void {
    expect(document.activeElement).toBe(element.getDOMNode());
}
