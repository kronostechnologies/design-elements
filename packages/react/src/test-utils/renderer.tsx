import { ThemeWrapped } from '@design-elements/test-utils/theme-wrapped';
import { mount, ReactWrapper, render } from 'enzyme';
import React, { Component, ReactElement } from 'react';
import { MemoryRouter } from 'react-router-dom';
import { DeviceContextProvider } from '../components/device-context-provider/device-context-provider';

export function mountWithProviders<C extends Component, P = C['props'], S = C['state']>(
    component: ReactElement<P>,
): ReactWrapper<P, S, C> {
    return mount(component, {
        wrappingComponent: AllProviders,
    });
}

export function renderWithProviders(
    component: ReactElement,
): Cheerio {
    return render(<AllProviders>{component}</AllProviders>);
}

function AllProviders({ children }: { children: ReactElement }): ReactElement {
    return (
        <MemoryRouter>
            <DeviceContextProvider>
                {ThemeWrapped(children)}
            </DeviceContextProvider>
        </MemoryRouter>
    );
}
