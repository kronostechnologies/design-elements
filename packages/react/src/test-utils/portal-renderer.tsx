import { render, RenderResult } from '@testing-library/react';
import React, { ReactElement } from 'react';
import { DeviceType } from '../components/device-context-provider/device-context-provider';
import { AllProviders } from './renderer';

export function renderPortalWithProviders(
    component: ReactElement,
    device?: DeviceType,
): RenderResult {
    return render(<AllProviders device={device}>{component}</AllProviders>);
}
