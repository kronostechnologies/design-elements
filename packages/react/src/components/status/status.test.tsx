import { Status, StatusTypes } from './status';
import { mountWithTheme } from '../../test-utils/renderer';

const statusTypesArray: StatusTypes[] = ['enabled', 'disabled', 'blocked'];

describe('Status', () => {
    statusTypesArray.forEach((type) => (
        it(`matches snapshot (${type})`, () => {
            const tree = mountWithTheme(<Status type={type} label={type} />);

            expect(tree).toMatchSnapshot();
        })
    ));
});
