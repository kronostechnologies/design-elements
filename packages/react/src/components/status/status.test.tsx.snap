// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Status matches snapshot (blocked) 1`] = `
.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c1 {
  background-color: #CD2C23;
  border: none;
  border-radius: 50%;
  box-sizing: border-box;
  height: 0.625rem;
  margin-right: var(--spacing-1halfx);
  width: 0.625rem;
}

<Status
  label="blocked"
  type="blocked"
>
  <styled.div
    type="blocked"
  >
    <div
      className="c0"
      type="blocked"
    >
      <styled.div
        type="blocked"
      >
        <div
          className="c1"
          type="blocked"
        />
      </styled.div>
      <span>
        blocked
      </span>
    </div>
  </styled.div>
</Status>
`;

exports[`Status matches snapshot (disabled) 1`] = `
.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  color: #B7BBC2;
}

.c1 {
  background-color: #F1F2F2;
  border: 1px solid #DBDEE1;
  border-radius: 50%;
  box-sizing: border-box;
  height: 0.625rem;
  margin-right: var(--spacing-1halfx);
  width: 0.625rem;
}

<Status
  label="disabled"
  type="disabled"
>
  <styled.div
    type="disabled"
  >
    <div
      className="c0"
      type="disabled"
    >
      <styled.div
        type="disabled"
      >
        <div
          className="c1"
          type="disabled"
        />
      </styled.div>
      <span>
        disabled
      </span>
    </div>
  </styled.div>
</Status>
`;

exports[`Status matches snapshot (enabled) 1`] = `
.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c1 {
  background-color: #008533;
  border: none;
  border-radius: 50%;
  box-sizing: border-box;
  height: 0.625rem;
  margin-right: var(--spacing-1halfx);
  width: 0.625rem;
}

<Status
  label="enabled"
  type="enabled"
>
  <styled.div
    type="enabled"
  >
    <div
      className="c0"
      type="enabled"
    >
      <styled.div
        type="enabled"
      >
        <div
          className="c1"
          type="enabled"
        />
      </styled.div>
      <span>
        enabled
      </span>
    </div>
  </styled.div>
</Status>
`;
